traffic_ops/traffic_ops_golang/api/generic_crud.go

```diff
@@ -118,28 +125,86 @@ func GenericCreateNameBasedID(val GenericCreator) (error, error, int) {
 	return nil, nil, http.StatusOK
 }
 
-func GenericRead(val GenericReader) ([]interface{}, error, error, int) {
+func MakeFirstQuery(val GenericReader, h map[string][]string, where string, orderBy string, pagination string, queryValues map[string]interface{}) (bool, time.Time) {
+	var max time.Time
+	ims := []string{}
+	runSecond := true
+	if h == nil {
+		return runSecond, max
+	}
+	ims = h[rfc.IfModifiedSince]
+	if len(ims) == 0 {
+		return runSecond, max
+	}
+	if l, ok := web.ParseHTTPDate(ims[0]); !ok {
+		log.Warnf("Date not parsable %v", ims[0])
+		return runSecond, max
+	} else {
+		query := val.SelectMaxLastUpdatedQuery(where, orderBy, pagination, val.GetType())
+		rows, err := val.APIInfo().Tx.NamedQuery(query, queryValues)
+		defer rows.Close()
+		if err != nil {
+			log.Warnf("Couldn't get the max last updated time: %v", err)
+			return runSecond, max
+		}
+		if err == sql.ErrNoRows {
+			runSecond = false
+			return runSecond, max
+		}
+		// This should only ever contain one row
+		if rows.Next() {
+			v := &ims2.LatestTimestamp{}
+			if err = rows.StructScan(v); err != nil || v == nil {
+				log.Warnf("Failed to parse the max time stamp into a struct %v", err)
+				return runSecond, max
+			}
+			max = v.LatestTime.Time
+			// The request IMS time is later than the max of (lastUpdated, deleted_time)
+			if l.After(v.LatestTime.Time) {
+				runSecond = false
+				return runSecond, max
+			}
+		} else {
+			runSecond = false
+		}
+	}
+	return runSecond, max
+}
+
+func GenericRead(h http.Header, val GenericReader, useIMS bool) ([]interface{}, error, error, int, *time.Time) {
+	vals := []interface{}{}
+	code := http.StatusOK
 	where, orderBy, pagination, queryValues, errs := dbhelpers.BuildWhereAndOrderByAndPagination(val.APIInfo().Params, val.ParamColumns())
 	if len(errs) > 0 {
-		return nil, util.JoinErrs(errs), nil, http.StatusBadRequest
+		return nil, util.JoinErrs(errs), nil, http.StatusBadRequest, nil
 	}
-
+	runSecond, maxTime := MakeFirstQuery(val, h, where, orderBy, pagination, queryValues)
+	if useIMS {
+		if !runSecond {
+			log.Debugln("IMS HIT")
```

Comment:
oh, by logging IMS hits vs misses, I was thinking it should be in the TO access logs, similar to the ATS cache result codes: https://docs.trafficserver.apache.org/en/8.0.x/admin-guide/logging/cache-results.en.html#cache-result-codes

We could have IMS_HIT, IMS_MISS, and NON_IMS. Ideally, each result code would be documented (I'm not sure if the TO access log is documented already or not).